### ISSUE TITLE
support `shouldOutputInFlatList` for menuItems

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#534]
@blocka made a great suggestion that menu items should be queryable in a flat list, allowing the client to determine how to output them in a hierarchy. (see https://github.com/wp-graphql/wp-graphql/issues/505#issuecomment-426654881)

The same discussion was had about hierarchical terms (categories, etc), and we implemented an argument `shouldOutputInFlatList` that when set to true will output all items in a non-hierarchical list and allow the client to build the hierarchy however they chose. 

See my comments here: https://github.com/wp-graphql/wp-graphql/issues/505#issuecomment-426654881

**Comment by jasonbahl** [comment]
Thinking more about this, we should probably support this behavior for _all_ Post Type Connections too 🤔 

Like if I were querying a hierarchical Post Type, I may want to get all items, regardless of their hierarchical level in a flat list, then do what I want with them in the client

**Comment by algebrathefish** [comment]
How would I output as a flat list? Where does the parameter go?

I have nested menu items and going through the same as posted here https://github.com/wp-graphql/wp-graphql/issues/505

**Comment by jasonbahl** [comment]
closed by #1348 

Menus and MenuItems are now output in a Flat list by defualt. 


### Query 
Take the following query for example: 

```graphql
query MENU_ITEMS {
  menuItems(where:{location:PRIMARY}) {
    nodes {
     key: id 
     parentId
     title: label
     url
    }
  }
}
```

### Query Results
This will return a flat list like so: 

```json
{
  "nodes": [
    {
      "key": "bmF2X21lbnVfaXRlbTo1NTE=",
      "parentId": null,
      "title": "test",
      "url": "http://acf2.local/test/",
    },
    {
      "key": "bmF2X21lbnVfaXRlbTo1NTI=",
      "parentId": null,
      "title": "Test Page",
      "url": "http://acf2.local/test-page/",
    },
    {
      "key": "bmF2X21lbnVfaXRlbTo2Mjk=",
      "parentId": null,
      "title": "Sample Page",
      "url": "http://acf2.local/sample-page/",
    },
    {
      "key": "bmF2X21lbnVfaXRlbTo2MzA=",
      "parentId": "bmF2X21lbnVfaXRlbTo2Mjk=",
      "title": "Child Page",
      "url": "http://acf2.local/sample-page/child-page/",
    },
    {
      "key": "bmF2X21lbnVfaXRlbTo2MzE=",
      "parentId": "bmF2X21lbnVfaXRlbTo2MzA=",
      "title": "Grandchild",
      "url": "http://acf2.local/sample-page/child-page/grandchild/",
      "__typename": "MenuItem"
    }
  ]
}
```

### Convert to hierarchical structure on the client

We can use this flat list to build a hierarchical structure on the client with a function like the following: 

```js
const flatListToHierarchical = (
    data = [],
    {idKey='key',parentKey='parentId',childrenKey='children'} = {}
) => {
    const tree = [];
    const childrenOf = {};
    data.forEach((item) => {
        const newItem = {...item};
        const { [idKey]: id, [parentKey]: parentId = 0 } = newItem;
        childrenOf[id] = childrenOf[id] || [];
        newItem[childrenKey] = childrenOf[id];
        parentId
            ? (
                childrenOf[parentId] = childrenOf[parentId] || []
            ).push(newItem)
            : tree.push(newItem);
    });
    return tree;
};
```

Passing the flat list through this will create the following: 

```json
[
  {
    "key": "bmF2X21lbnVfaXRlbTo1NTE=",
    "parentId": null,
    "title": "test",
    "url": "http://acf2.local/test/",
    "children": []
  },
  {
    "key": "bmF2X21lbnVfaXRlbTo1NTI=",
    "parentId": null,
    "title": "Test Page",
    "url": "http://acf2.local/test-page/",
    "children": []
  },
  {
    "key": "bmF2X21lbnVfaXRlbTo2Mjk=",
    "parentId": null,
    "title": "Sample Page",
    "url": "http://acf2.local/sample-page/",
    "children": [
      {
        "key": "bmF2X21lbnVfaXRlbTo2MzA=",
        "parentId": "bmF2X21lbnVfaXRlbTo2Mjk=",
        "title": "Child Page",
        "url": "http://acf2.local/sample-page/child-page/",
        "children": [
          {
            "key": "bmF2X21lbnVfaXRlbTo2MzE=",
            "parentId": "bmF2X21lbnVfaXRlbTo2MzA=",
            "title": "Grandchild",
            "url": "http://acf2.local/sample-page/child-page/grandchild/",
            "children": []
          }
        ]
      }
    ]
  }
]
```

And this allows us to efficiently query the data in a flat list, while also using in hierarchical components such as a Tree View component or nested Menus. 

### Flat List in Ant Design Tree Component

For example:

```js
const MENU_QUERY = gql`
query MENU_ITEMS($location:MenuLocationEnum){
  menuItems(first:100 where:{location:$location}) {
    nodes {
     key: id 
     parentId
     title: label
     url
    }
  }
}
`;

const TreeExample = () => {
    ## Using Apollo
    const { loading, error, data } = useQuery(MENU_QUERY, {
        variables: {
            location: "PRIMARY"
        }
    });
    if (loading) return <p>Loading...</p>;
    if (error) return <p>Error :(</p>;
    const { menuItems } = data;
    const menu = flatListToHierarchical(menuItems.nodes);
    return (
        <Tree
            checkable
            treeData={menu}
        />
    );
};
```

Outputs this component: 

![Screen Shot 2020-06-23 at 9 22 38 AM](https://user-images.githubusercontent.com/1260765/85445462-a779a600-b550-11ea-8484-255ab273240a.png)

### Flat List in Ant Design Menu Component

This isn't the prettiest code or best way to to this, just proof of concept that flat lists can hydrate hierarchical components.

```js
const MenuExample = () => {
    const { loading, error, data } = useQuery(MENU_QUERY, {
        variables: {
            location: "PRIMARY"
        }
    });
    if (loading) return <p>Loading...</p>;
    if (error) return <p>Error :(</p>;
    const { menuItems } = data;
    const menu = flatListToHierarchical(menuItems.nodes);
    return(
        <Menu mode="horizontal">
            {menu.map((item => {
                return item.children && item.children.length ? (
                    <SubMenu key={item.key} title={item.title}>
                        {item.children.map(child => {
                            return(
                                <Menu.Item key={child.key}>{child.title}</Menu.Item>
                            );
                        })}

                    </SubMenu>
                    ) : (
                    <Menu.Item key={item.key} >{item.title}</Menu.Item>
                );
            }))}
        </Menu>
    );

};
```

Outputs the following:

![menu-gif](https://user-images.githubusercontent.com/1260765/85445784-fcb5b780-b550-11ea-85ed-2ef41af1c985.gif)
